tensorboard/db.py

```diff
@@ -65,39 +88,42 @@ def get_plugin_ids(self, names):
     :type names: list[str]
     :rtype: dict[str, int]
     """
-    result = {}
-    names_set = set(names)
     with self._plugin_ids_by_name_lock:
-      for name in names:
-        id_ = self._plugin_ids_by_name.get(name)
-        if id_ is not None:
-          result[name] = id_
-          names_set.remove(name)
-      if names_set:
-        # TODO(@jart): Retry transaction with exponential backoff.
-        with contextlib.closing(self._db_connection_provider()) as conn:
-          with conn:
-            with contextlib.closing(conn.cursor()) as c:
-              c.execute('SELECT plugin_id, name FROM Plugins')
-              self._plugin_ids_by_name.clear()
-              max_id = 0
-              for id_, name in c.fetchall():
-                if id_ > max_id:
-                  max_id = id_
-                self._plugin_ids_by_name[name] = id_
-                if name in names_set:
-                  result[name] = id_
-                  names_set.remove(name)
-              new_rows = []
-              for name in names_set:
-                max_id += 1
-                result[name] = max_id
-                new_rows.append((max_id, name))
-              if new_rows:
-                c.executemany(
-                    'INSERT INTO Plugins (plugin_id, name) VALUES (?, ?)',
-                    new_rows)
-    return result
+      if all(name in self._plugin_ids_by_name for name in names):
+        return self._get_plugin_ids(names)
+      self._plugin_ids_by_name.update(
+          self.run_transaction(functools.partial(_add_plugins, names)))
```

Comment:
The `discard` method doesn't stand out, but that's what's doing the trick. I should also rename this function too.